File: src/libsemigroups/transf.jl
# Copyright (c) 2026, James W. Swent, J. D. Mitchell
#
# Distributed under the terms of the GPL license version 3.
#
# The full license is in the file LICENSE, distributed with this software.

"""
transf.jl - Julia wrappers for libsemigroups transformation classes

This file provides low-level Julia wrappers for the C++ transformation
classes (Transf, PPerm, Perm) exposed via CxxWrap. These are thin wrappers
that provide memory safety via GC.@preserve and convenient access to the
C++ bindings.
"""

# ============================================================================
# Type aliases - Concrete template instantiations from C++
# ============================================================================

"""
    Transf1

Transformation with uint8_t scalar type (supports degrees up to 2^8 = 256).
This is the underlying C++ type Transf<0, uint8_t>.
"""
const Transf1 = LibSemigroups.Transf1

"""
    Transf2

Transformation with uint16_t scalar type (supports degrees up to 2^16 = 65536).
This is the underlying C++ type Transf<0, uint16_t>.
"""
const Transf2 = LibSemigroups.Transf2

"""
    Transf4

Transformation with uint32_t scalar type (supports degrees up to 2^32).
This is the underlying C++ type Transf<0, uint32_t>.
"""
const Transf4 = LibSemigroups.Transf4

"""
    PPerm1

Partial permutation with uint8_t scalar type (supports degrees up to 2^8 = 256).
This is the underlying C++ type PPerm<0, uint8_t>.
"""
const PPerm1 = LibSemigroups.PPerm1

"""
    PPerm2

Partial permutation with uint16_t scalar type (supports degrees up to 2^16 = 65536).
This is the underlying C++ type PPerm<0, uint16_t>.
"""
const PPerm2 = LibSemigroups.PPerm2

"""
    PPerm4

Partial permutation with uint32_t scalar type (supports degrees up to 2^32).
This is the underlying C++ type PPerm<0, uint32_t>.
"""
const PPerm4 = LibSemigroups.PPerm4

"""
    Perm1

Permutation with uint8_t scalar type (supports degrees up to 2^8 = 256).
This is the underlying C++ type Perm<0, uint8_t>.
"""
const Perm1 = LibSemigroups.Perm1

"""
    Perm2

Permutation with uint16_t scalar type (supports degrees up to 2^16 = 65536).
This is the underlying C++ type Perm<0, uint16_t>.
"""
const Perm2 = LibSemigroups.Perm2

"""
    Perm4

Permutation with uint32_t scalar type (supports degrees up to 2^32).
This is the underlying C++ type Perm<0, uint32_t>.
"""
const Perm4 = LibSemigroups.Perm4

# Union types for generic functions (for internal use in this file)
const _TransfTypes = Union{Transf1,Transf2,Transf4}
const _PPermTypes = Union{PPerm1,PPerm2,PPerm4}
const _PermTypes = Union{Perm1,Perm2,Perm4}
const _PTransfTypes = Union{_TransfTypes,_PPermTypes,_PermTypes}

# ============================================================================
# Instance methods - Direct wrappers
# ============================================================================
degree(t::_PTransfTypes) = Int(LibSemigroups.degree(t))

rank(t::_PTransfTypes) = Int(LibSemigroups.rank(t))

hash_value(t::_PTransfTypes) = LibSemigroups.hash(t)

images_vector(t::_PTransfTypes) = LibSemigroups.images_vector(t)

# ============================================================================
# Methods requiring GC.@preserve for memory safety
# ============================================================================
function product_inplace!(result::T, x::T, y::T) where {T<:_PTransfTypes}
    GC.@preserve result x y begin
        @wrap_libsemigroups_call LibSemigroups.product_inplace!(result, x, y)
    end
    return nothing
end

function increase_degree_by!(t::T, n::Integer) where {T<:_PTransfTypes}
    GC.@preserve t begin
        @wrap_libsemigroups_call LibSemigroups.increase_degree_by!(t, UInt(n))
    end
    return t
end

function swap!(t::T, x::T) where {T<:_PTransfTypes}
    GC.@preserve t x begin
        @wrap_libsemigroups_call LibSemigroups.swap(t, x)
    end
    return nothing
end

function copy(t::_PTransfTypes)
    return LibSemigroups.copy(t)
end

# ============================================================================
# Module-level helper functions
# ============================================================================

# Return the identity transformation with the same degree as `t`.
one(t::_PTransfTypes) = LibSemigroups.one(t)

# Return the identity transformation of type `T` with degree `n`.
one(::Type{Transf1}, n::Integer) = LibSemigroups.one(Transf1, UInt(n))
one(::Type{Transf2}, n::Integer) = LibSemigroups.one(Transf2, UInt(n))
one(::Type{Transf4}, n::Integer) = LibSemigroups.one(Transf4, UInt(n))
one(::Type{PPerm1}, n::Integer) = LibSemigroups.one(PPerm1, UInt(n))
one(::Type{PPerm2}, n::Integer) = LibSemigroups.one(PPerm2, UInt(n))
one(::Type{PPerm4}, n::Integer) = LibSemigroups.one(PPerm4, UInt(n))
one(::Type{Perm1}, n::Integer) = LibSemigroups.one(Perm1, UInt(n))
one(::Type{Perm2}, n::Integer) = LibSemigroups.one(Perm2, UInt(n))
one(::Type{Perm4}, n::Integer) = LibSemigroups.one(Perm4, UInt(n))

image(t::_PTransfTypes) = LibSemigroups.image(t)

domain(t::_PTransfTypes) = LibSemigroups.domain(t)

inverse(p::Union{_PPermTypes,_PermTypes}) = LibSemigroups.inverse(p)

left_one(p::_PPermTypes) = LibSemigroups.left_one(p)

right_one(p::_PPermTypes) = LibSemigroups.right_one(p)

File: deps/src/transf.cpp
//
// Semigroups.jl
// Copyright (C) 2026, James W. Swent
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include "libsemigroups_julia.hpp"
#include <libsemigroups/transf.hpp>

namespace libsemigroups_julia {

namespace {

template <typename PTransfType>
void bind_ptransf_common(jl::Module &                      m,
                         jlcxx::TypeWrapper<PTransfType> & type,
                         std::string const &               type_name)
{
  using Scalar = typename PTransfType::point_type;

  ////////////////////////////////////////////////////////////////////////
  // Constructor
  ////////////////////////////////////////////////////////////////////////

  // Safe constructor using make<> which validates inputs
  // Julia passes 1-based indices (0 = UNDEFINED for PPerm)
  m.method(type_name, [](std::vector<Scalar> const & imgs) -> PTransfType {
    return libsemigroups::make<PTransfType>(vec_to_0_based_undef(imgs));
  });

  ////////////////////////////////////////////////////////////////////////
  // Special methods (Python __xxx__ equivalents)
  ////////////////////////////////////////////////////////////////////////

  // copy / __copy__ - explicit copy constructor
  type.method("copy", [](PTransfType const & self) {
    return PTransfType(self);
  });

  // __getitem__ equivalent - element access with bounds checking
  // Julia passes 1-based index, returns 1-based result (0 = UNDEFINED)
  type.method("getindex", [](PTransfType const & self, size_t i) -> Scalar {
    return to_1_based_undef(self.at(to_0_based(i)));
  });

  // __hash__ equivalent
  type.method("hash", &PTransfType::hash_value);

  ////////////////////////////////////////////////////////////////////////
  // Instance methods
  ////////////////////////////////////////////////////////////////////////

  type.method("degree", &PTransfType::degree);
  type.method("rank", &PTransfType::rank);

  // swap - use lambda to avoid exposing PTransfBase in signature
  type.method("swap", [](PTransfType & self, PTransfType & other) {
    self.swap(other);
  });

  // increase_degree_by - modifies in place, don't expose base class return type
  type.method("increase_degree_by!", [](PTransfType & self, size_t m) {
    self.increase_degree_by(m);
  });

  // images - return a Julia-compatible vector with 1-based indices
  type.method("images_vector", [](PTransfType const & self) {
    std::vector<Scalar> result;
    result.reserve(self.degree());
    for (auto val : self)
    {
      result.push_back(to_1_based_undef(val));
    }
    return result;
  });

  ////////////////////////////////////////////////////////////////////////
  // Static methods
  ////////////////////////////////////////////////////////////////////////

  // one(n) - static method that creates identity of degree n
  m.method("one", [](jlcxx::SingletonType<PTransfType>, size_t n) {
    return PTransfType::one(n);
  });

  ////////////////////////////////////////////////////////////////////////
  // Comparison operators
  ////////////////////////////////////////////////////////////////////////

  // NOTE: We use named functions as CxxWrap doesn't properly bind
  // operator symbols as Julia-callable methods

  type.method("is_equal", [](PTransfType const & a, PTransfType const & b) -> bool {
    return a == b;
  });

  type.method("is_not_equal", [](PTransfType const & a, PTransfType const & b) -> bool {
    return a != b;
  });

  type.method("is_less", [](PTransfType const & a, PTransfType const & b) -> bool {
    return a < b;
  });

  type.method("is_less_equal", [](PTransfType const & a, PTransfType const & b) -> bool {
    return a <= b;
  });

  type.method("is_greater", [](PTransfType const & a, PTransfType const & b) -> bool {
    return a > b;
  });

  type.method("is_greater_equal",
              [](PTransfType const & a, PTransfType const & b) -> bool {
                return a >= b;
              });

  type.method("multiply", [](PTransfType const & a, PTransfType const & b) {
    return a * b;
  });

  m.method("product_inplace!",
           [](PTransfType & xy, PTransfType const & x, PTransfType const & y) {
    xy.product_inplace(x, y);
  });
}

template <typename TransfType>
void bind_transf_type(jl::Module & m, std::string const & name)
{
  auto type = m.add_type<TransfType>(name);
  bind_ptransf_common(m, type, name);
}

template <typename PPermType>
void bind_pperm_type(jl::Module & m, std::string const & name)
{
  using Scalar = typename PPermType::point_type;

  auto type = m.add_type<PPermType>(name);
  bind_ptransf_common(m, type, name);

  // Constructor from domain, image, and degree using make<>
  // Julia passes 1-based domain/image vectors, degree is a count (no conversion)
  m.method(name,
           [](std::vector<Scalar> const & dom, std::vector<Scalar> const & img,
              size_t deg) -> PPermType {
             return libsemigroups::make<PPermType>(vec_to_0_based(dom),
                                                   vec_to_0_based(img), deg);
           });
}

template <typename PermType> void bind_perm_type(jl::Module & m, std::string const & name)
{
  auto type = m.add_type<PermType>(name);
  bind_ptransf_common(m, type, name);
}

}    // anonymous namespace

// Main function to define all transformation bindings
void define_transf(jl::Module & m)
{
  using namespace libsemigroups;

  ////////////////////////////////////////////////////////////////////////
  // Bind concrete instantiations: Transf
  ////////////////////////////////////////////////////////////////////////

  bind_transf_type<Transf<0, uint8_t>>(m, "Transf1");
  bind_transf_type<Transf<0, uint16_t>>(m, "Transf2");
  bind_transf_type<Transf<0, uint32_t>>(m, "Transf4");

  ////////////////////////////////////////////////////////////////////////
  // Bind concrete instantiations: PPerm
  ////////////////////////////////////////////////////////////////////////

  bind_pperm_type<PPerm<0, uint8_t>>(m, "PPerm1");
  bind_pperm_type<PPerm<0, uint16_t>>(m, "PPerm2");
  bind_pperm_type<PPerm<0, uint32_t>>(m, "PPerm4");

  ////////////////////////////////////////////////////////////////////////
  // Bind concrete instantiations: Perm
  ////////////////////////////////////////////////////////////////////////

  bind_perm_type<Perm<0, uint8_t>>(m, "Perm1");
  bind_perm_type<Perm<0, uint16_t>>(m, "Perm2");
  bind_perm_type<Perm<0, uint32_t>>(m, "Perm4");

  ////////////////////////////////////////////////////////////////////////
  // Module-level helper functions: one
  ////////////////////////////////////////////////////////////////////////

  // one(f) - returns identity of same degree as f
  m.method("one", [](Transf<0, uint8_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](Transf<0, uint16_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](Transf<0, uint32_t> const & f) {
    return libsemigroups::one(f);
  });

  m.method("one", [](PPerm<0, uint8_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](PPerm<0, uint16_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](PPerm<0, uint32_t> const & f) {
    return libsemigroups::one(f);
  });

  m.method("one", [](Perm<0, uint8_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](Perm<0, uint16_t> const & f) {
    return libsemigroups::one(f);
  });
  m.method("one", [](Perm<0, uint32_t> const & f) {
    return libsemigroups::one(f);
  });

  ////////////////////////////////////////////////////////////////////////
  // Module-level helper functions: image
  ////////////////////////////////////////////////////////////////////////

  // image(f) - returns sorted vector of image points (1-based, excluding UNDEFINED)
  m.method("image", [](Transf<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](Transf<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](Transf<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });

  m.method("image", [](PPerm<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](PPerm<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](PPerm<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });

  m.method("image", [](Perm<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](Perm<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });
  m.method("image", [](Perm<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::image(f));
  });

  ////////////////////////////////////////////////////////////////////////
  // Module-level helper functions: domain
  ////////////////////////////////////////////////////////////////////////

  // domain(f) - returns sorted vector of domain points (1-based)
  m.method("domain", [](Transf<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](Transf<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](Transf<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });

  m.method("domain", [](PPerm<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](PPerm<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](PPerm<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });

  m.method("domain", [](Perm<0, uint8_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](Perm<0, uint16_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });
  m.method("domain", [](Perm<0, uint32_t> const & f) {
    return vec_to_1_based(libsemigroups::domain(f));
  });

  ////////////////////////////////////////////////////////////////////////
  // Module-level helper functions: inverse (PPerm and Perm only)
  ////////////////////////////////////////////////////////////////////////

  // inverse(f) - returns inverse partial perm or perm
  m.method("inverse", [](PPerm<0, uint8_t> const & f) {
    return libsemigroups::inverse(f);
  });
  m.method("inverse", [](PPerm<0, uint16_t> const & f) {
    return libsemigroups::inverse(f);
  });
  m.method("inverse", [](PPerm<0, uint32_t> const & f) {
    return libsemigroups::inverse(f);
  });

  m.method("inverse", [](Perm<0, uint8_t> const & f) {
    return libsemigroups::inverse(f);
  });
  m.method("inverse", [](Perm<0, uint16_t> const & f) {
    return libsemigroups::inverse(f);
  });
  m.method("inverse", [](Perm<0, uint32_t> const & f) {
    return libsemigroups::inverse(f);
  });

  ////////////////////////////////////////////////////////////////////////
  // Module-level helper functions: left_one and right_one (PPerm only)
  ////////////////////////////////////////////////////////////////////////

  // left_one(f) - partial perm that fixes domain points
  m.method("left_one", [](PPerm<0, uint8_t> const & f) {
    return libsemigroups::left_one(f);
  });
  m.method("left_one", [](PPerm<0, uint16_t> const & f) {
    return libsemigroups::left_one(f);
  });
  m.method("left_one", [](PPerm<0, uint32_t> const & f) {
    return libsemigroups::left_one(f);
  });

  // right_one(f) - partial perm that fixes image points
  m.method("right_one", [](PPerm<0, uint8_t> const & f) {
    return libsemigroups::right_one(f);
  });
  m.method("right_one", [](PPerm<0, uint16_t> const & f) {
    return libsemigroups::right_one(f);
  });
  m.method("right_one", [](PPerm<0, uint32_t> const & f) {
    return libsemigroups::right_one(f);
  });
}

}    // namespace libsemigroups_julia
